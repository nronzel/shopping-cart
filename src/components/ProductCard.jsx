import {
  Card,
  CardBody,
  CardFooter,
  Image,
  Heading,
  Divider,
  ButtonGroup,
  Button,
  Stack,
  Text,
  useColorModeValue,
} from "@chakra-ui/react";
import React from "react";

const ProductCard = ({ productTitle, imgSource, price }) => {
  return (
    <Card maxW="xs" borderRadius="xl" bgColor={useColorModeValue("gray.100", "gray.700")}>
      <CardBody>
        <Image
          h="250px"
          w="100%"
          objectFit="cover"
          objectPosition="center"
          src={imgSource}
          borderRadius="lg"
          alt="product image"
        />
        <Stack spacing={3} mt={4}>
          <Heading size="md">{productTitle}</Heading>
          <Text fontWeight="thin">
            Lorem ipsum, dolor sit amet consectetur adipisicing elit. Dolor in
            dolorem fugit quo obcaecati facere!
          </Text>
          <Text
            color={useColorModeValue("pink.500", "pink.200")}
            fontSize="xl"
            fontWeight="bold"
          >
            {price}
          </Text>
        </Stack>
      </CardBody>
      <Divider />
      <CardFooter>
        <ButtonGroup>
          <Button variant="solid" colorScheme="pink">Add to cart</Button>
          <Button variant="outline" fontWeight="normal">
            Details
          </Button>
        </ButtonGroup>
      </CardFooter>
    </Card>
  );
};

export default ProductCard;
