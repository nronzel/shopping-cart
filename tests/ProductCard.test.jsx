import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";
import { ChakraProvider } from "@chakra-ui/react";
import React from "react";
import ProductCard from "../src/components/ProductCard";

const product = [
  {
    id: 0,
    name: "product 1",
    price: 75,
    src: "../public/products/product1.jpg",
  },
];

describe("Product card renders", () => {
  it("renders product name", () => {
    render(
      <ChakraProvider>
        <ProductCard product={product[0]} />
      </ChakraProvider>
    );
    expect(screen.getByText(/product 1/i)).toBeInTheDocument();
  });

  it("renders product price", () => {
    render(
      <ChakraProvider>
        <ProductCard product={product[0]} />
      </ChakraProvider>
    );
    expect(screen.getByText(/75/i)).toBeInTheDocument();
  });

  it("renders the correct image alt text", () => {
    render(
      <ChakraProvider>
        <ProductCard product={product[0]} />
      </ChakraProvider>
    );
    const displayedImage = document.querySelector("img");
    expect(displayedImage.alt).toContain("product image");
  });

  it("renders the correct image source", () => {
    render(
      <ChakraProvider>
        <ProductCard product={product[0]} />
      </ChakraProvider>
    );
    const displayedImage = document.querySelector("img");
    expect(displayedImage.src).toContain("/products/product1.jpg");
  });

  it("renders add to cart button", () => {
    render(
      <ChakraProvider>
        <ProductCard product={product[0]} />
      </ChakraProvider>
    );
    expect(screen.getByText(/add to cart/i)).toBeInTheDocument();
  });

  it("renders details button", () => {
    render(
      <ChakraProvider>
        <ProductCard product={product[0]} />
      </ChakraProvider>
    );
    expect(screen.getByText(/details/i)).toBeInTheDocument();
  });

  it("add to cart button calls addToCart function once when clicked", () => {
    const addToCart = vi.fn();
    const { getByRole } = render(
      <ChakraProvider>
        <ProductCard product={product[0]} addToCart={addToCart} />
      </ChakraProvider>
    );

    const addBtn = getByRole("button", { name: "add to cart" });

    fireEvent.click(addBtn);

    expect(addToCart).toHaveBeenCalledWith(product[0]);
    expect(addToCart).toHaveBeenCalledTimes(1);
  });

  it("matches the snapshot", () => {
    const { container } = render(
      <ChakraProvider>
        <ProductCard product={product[0]} />
      </ChakraProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
