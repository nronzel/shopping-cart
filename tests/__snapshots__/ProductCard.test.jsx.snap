// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product card renders > matches the snapshot 1`] = `
<div>
  <div
    class="chakra-card css-m47irm"
  >
    <div
      class="chakra-card__body css-1idwstw"
    >
      <img
        alt="product image"
        class="chakra-image css-aucjl3"
        src="/public/products/product1.jpg"
      />
      <div
        class="chakra-stack css-t7aooi"
      >
        <h2
          class="chakra-heading css-18j379d"
        >
          product 1
        </h2>
        <p
          class="chakra-text css-eefv6g"
        >
          Lorem ipsum, dolor sit amet consectetur adipisicing elit. Dolor in dolorem fugit quo obcaecati facere!
        </p>
        <p
          class="chakra-text css-1bpuy0u"
        >
          75
        </p>
      </div>
    </div>
    <hr
      aria-orientation="horizontal"
      class="chakra-divider css-svjswr"
    />
    <div
      class="chakra-card__footer css-1a7q9t8"
    >
      <div
        class="chakra-button__group css-1ind3v2"
        data-orientation="horizontal"
        role="group"
      >
        <button
          class="chakra-button css-9qq7ve"
          type="button"
        >
          Add to cart
        </button>
        <button
          class="chakra-button css-11kckzb"
          type="button"
        >
          Details
        </button>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
